specs/ChainSpec/Upgradability.md

```diff
@@ -0,0 +1,71 @@
+# Upgradability
+
+This part of specification describes specifics of upgrading the protocol, and touches on few different parts of the system.
+
+High level upgradability consists of three parts:
+1. Updating without any changes to underlaying data structures or protocol;
+2. Updating when underlaying data structures changed (config, database or something else internal to the node and probably client specific);
+3. Updating with protocol changes that all validating nodes must adjust to.
+
+## Versioning
+
+There are 2 different important versions:
+- Version of binary defines it's internal data structures / database and configs. This version is client specific and doesn't need to be matching between nodes.
+- Version of the protocol, defining the "language" nodes are speaking.
+
+```rust
+/// Latest version of protocol that this binary can work with.
+type ProtocolVersion = u32;
+```
+
+Open questions: do we want to use something better than a number of versioning protocol?
+Ideally it should be some hash of protocol spec that this binary supports, but we don't have a canonical protocol spec versioning yet. We can use commit hash of `spec` / `NEPs` repo.
```

Comment:
The nice property with numeric values is that you can easily compare them to see which one is newer. With hash it becomes much more difficult.

Reply:
The version order requires consensus by itself. In the decentralized setting version ordering can be only determined by order of their acceptance in the blockchain.

Reply:
If it's the commit hash, we have git history to compare, it's okay for scripts but more difficult human. What about
- a web page for listing spec hash <> spec
- or use a combination of <number>_<hash>